### ISSUE TITLE
Using bt headset and listening music same time

### ISSUE THREAD

**Issue by Lewik** [mumble-voip/mumble#4062]
**Context**
Using bt headset and listening music same time

**Describe the feature you have in mind**
Switch between HFP/HSP and A2DP when PTT is setted up

**Describe alternatives you've considered**
 

**Additional context**

--------------------

As I know (not sure *) windows will use one protocol at one time: either HFP/HSP (phones with good sound, no mic)  or A2DP (a headset with mic and poor sound).

The temporary solution is to use HFP/HSP and wired mic.

As I know, android can't use both protocols too, but there is no trouble to make a call when listening to music. You will not hear music during call dut you don`t need to make any additional moves to make a call. Look like it just switches protocols.

So the question is: could mumble switch between protocols when PTT option is enabled?

* For example https://stackoverflow.com/questions/40951060/how-to-make-a2dp-and-hsp-work-simultaneously

**Comment by Krzmbrzl** [comment]
So what you'd want is to use HFP/HSP when not talking and automatically switch to A2DP when pressing the PTT button and once you release it, back again?

**Comment by Lewik** [comment]
@Krzmbrzl Yes. Correct.

At least this will satisfy users that use 100% attenuation.
If you know mumble alternative who did that already - tell us) (We love mumble anyway)

**Comment by Krzmbrzl** [comment]
Tbh I'm not sure this is Mumble's job to do. And from what I have found out in a very quick research, it isn't something that is easily doable either.

I could also imagine that switching the Bluetooth profile repeatedly could cause all sorts of problems (even if Mumble itself deals well with it).

TL;DR: I think the chosing of the BT profile is up to the user and non of Mumble's business.

I'll check with the rest of the team though. Maybe they see that differently :)

**Comment by Lewik** [comment]
@Krzmbrzl 
Ok. Another approach.
Given I am listening to music. Than starting mumble. Music disappears. When I am closing mumble - music on again. I am not a windows/desktop programmer but it seems when mumble starts and stops it somehow occupies sound device. I hope you do it by your self with come API calls and can handle this occupation while PTT. In my case, these switches are enough fast to use in PTT.

**Comment by Krzmbrzl** [comment]
I don't know the details of the audio implementation we use but I think the audio backend is responsible for doing this kind of stuff during its initialization and its shutdown.
Doing this every time you push and release the PTT button doesn't make any sense though.

Plus I also have the feeling that this is Windows-specfic and not needed/doable on other OS. Thus even if we decided to extract this functionality to some API that can be triggered via the PTT button, it would mean a lot if work. And in the end I just don't think it is worth it (and also if things are as I think they are, this kind of stuff would clutter our code base unnecessarily with a lot of case-differentiation for Windows and non-Windows)...

**Comment by Lewik** [comment]
@Krzmbrzl 
> Plus I also have the feeling that this is Windows-specfic and not needed/doable on other OS.

I think it's Bluetooth specific. Channel bandwidth is not enough. If it was enough there were be a protocol for good sound and voice.

**Comment by davidebeatrici** [comment]
We don't handle Bluetooth directly, we talk to audio devices through WASAPI.

Music playback probably stops because the audio device changes (stereo -> mono), restarting the player should fix it.

Stopping the audio input engine should indeed make Windows switch to HFP/HSP mode, unless other applications are using the microphone.

We could add a setting to do that automatically for PTT mode, however I'm not sure whether it's feasible in terms of performance.

**Comment by Krzmbrzl** [comment]
> We could add a setting to do that automatically for PTT mode, however I'm not sure whether it's feasible in terms of performance.

Yep that's what I was thinking. If we were to continuously start and stop the audio input engine, that'd probably produce a lot of overhead for (at least in my opinion) comparatively little gain.

I think this is one of the things that if someone brought in a PR with this feature being implemented as an optional addition, we'd probably merge it in. However I just don't see this as something we'd be implementing by ourselves.

So if you (or someone seeing this) want to give this a shot, you're very welcome to do so, but for now I'll close the issue.

If you feel like we/I have missed important bits, please feel free to point them out :)
